source/server/listener_impl.cc

```diff
@@ -668,9 +668,11 @@ void ListenerImpl::buildSocketOptions() {
 #else
     // Not in place listener update.
     if (config_.has_connection_balance_config()) {
-      // Currently exact balance is the only supported type and there are no options.
-      ASSERT(config_.connection_balance_config().has_exact_balance());
-      connection_balancer_ = std::make_shared<Network::ExactConnectionBalancerImpl>();
+      if (config_.connection_balance_config().has_exact_balance()) {
+        connection_balancer_ = std::make_shared<Network::ExactConnectionBalancerImpl>();
+      } else if (config_.connection_balance_config().has_extend_balance()) {
+        // TODO(daixiang0) implement extend balancer.
```

Comment:
@mattklein123 I find that it will bring dependencies from `contrib` explicitly, do we need to create a factory to get connection balancer based on string name? If so, we do not introduce header files from `contrib` into `source`.

Reply:
There should be no dependencies on contrib from core. You will likely need to introduce a new extension point.

Reply:
I want to create a factory like socketInterface does:
https://github.com/envoyproxy/envoy/blob/569f23993160c01bd07af94f2d8e1bb973933c0a/source/server/server.cc#L578-L585

I will go in this direction if you are ok with it.

Reply:
@mattklein123 can I move in this direction?

Reply:
I'm not sure which direction you mean. If you mean adding an extension point that sounds r asonable but we will need to see the proposed changes.

Reply:
I mean create a factory as socketInterface does: 
https://github.com/envoyproxy/envoy/blob/569f23993160c01bd07af94f2d8e1bb973933c0a/source/server/server.cc#L578-L580

I will propose changes soon.

Reply:
@mattklein123 https://github.com/envoyproxy/envoy/pull/20268/commits/3dd53f2c19d450f0c4b679ab390e2b1575663150 shows my idea: introduce a factory, then any new connection balancer inherits `ConnectionBalancerBase`. In this way, we do not introduce any dependency from `contrib`. What do you think about this way?

There may be an issue that converts raw pointer to smart pointer, I am trying to fix it.